libnetwork/ipams/defaultipam/structures.go

```diff
@@ -59,13 +68,44 @@ func PoolIDFromString(str string) (pID PoolID, err error) {
 	return pID, nil
 }
 
+func parsePoolIDV2(str string) (pID PoolID, err error) {
+	str = strings.TrimSuffix(strings.TrimPrefix(str, "PoolID{"), "}")
+
+	for _, field := range strings.FieldsFunc(str, func(c rune) bool { return c == ';' }) {
+		p := strings.SplitN(field, "=", 2)
+		switch p[0] {
+		case "AddressSpace":
+			pID.AddressSpace = p[1]
+		case "Subnet":
+			if pID.Subnet, err = netip.ParsePrefix(p[1]); err != nil {
+				return PoolID{}, types.InvalidParameterErrorf("invalid string form for subnetkey: %s", str)
+			}
+		case "ChildSubnet":
+			if pID.ChildSubnet, err = netip.ParsePrefix(p[1]); err != nil {
+				return PoolID{}, types.InvalidParameterErrorf("invalid string form for subnetkey: %s", str)
+			}
+		}
```

Comment:
Is the plan to add more fields? ... this will silently drop fields it doesn't recognise, is that likely to be safe behaviour on downgrade?

I guess it's impossible to predict, but it might be safest to bomb out? Or, could come up with a naming convention that says whether the address space is safe to use without understanding a field?

(Could log unknown fields, but that might not help much.)

Reply:
> Is the plan to add more fields? ... this will silently drop fields it doesn't recognise, is that likely to be safe behaviour on downgrade?

Yeah, we might need a new field to properly support duplicate static allocations (to uniquely identify each duplicate).

> this will silently drop fields it doesn't recognise, is that likely to be safe behaviour on downgrade? I guess it's impossible to predict, but it might be safest to bomb out?

Indeed, my intention was to make it easy to add new fields while providing a safe / easy downgrade path.

> Or, could come up with a naming convention that says whether the address space is safe to use without understanding a field?

Hm not sure to see what you mean 🤔 We can't predict what future fields (if any) will look like. Each major release will understand a specific set of fields, matching a specific implementation of the Allocator / addrSpace.

> (Could log unknown fields, but that might not help much.)

We could log, yeah. But this would happen only on downgrade (so not really useful) or if we commit a big mistake (eg. a typo in a field name in the serialize code). Not sure if that's really valuable 🤔

Reply:
> Hm not sure to see what you mean 🤔 We can't predict what future fields (if any) will look like. Each major release will understand a specific set of fields, matching a specific implementation of the Allocator / addrSpace.

I was thinking there may be two categories of new field in the id ... informational, some annotation for debug/logging/inspect that old builds can safely ignore. Or, something new that means an old daemon can't safely use the pool because it doesn't know how it works.

The duplicate-tracking field is probably in the second category? I guess an old daemon that didn't understand the field would see duplicate pool-ids, maybe lose data from one of the pools, and generally be a bit weird? (So, we'd want it to fail if the new id-field is used. In which case, it's an example of a field that shouldn't be silently ignored.)

_For future backwards-compatibility - we could come up with a convention along the lines of "an unknown field name with prefix 'Info' can be silently ignored, other unknown fields mean the pool is unusable". We should probably make sure the daemon starts with unusable pools, so that they can be deleted and re-created. Maybe it's not necessary, I don't have examples, but this would be the time to build it in if it might be needed._